### ISSUE TITLE
Discussion: merge `main` to `production_testing`?

### ISSUE THREAD

**Issue by hanhou** [AllenNeuralDynamics/dynamic-foraging-task#97]
I have a question about git best practice @alexpiet @XX-Yin: should we always merge `main` to `production_testing` each time when we merge `production_testing` to `main`?

The reason is that if `production_testing` is not caught up with `main`, when one creates a new branch from `main` (in the screenshot below, `han-automatic-training` was created from `main` from point A) and then tries to issue a PR onto `production_testing`, the diff may include additional commits than what are really needed.

<img width="870" alt="image" src="https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/24734299/bcc359d5-28a3-4537-af4a-427452b0d9cf">

Here I just meant to include the two commits in red circle, but I got a very long diff that includes some old commits

<img width="1129" alt="image" src="https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/24734299/efd7e361-00ad-4d9b-8b08-550551e89ace">

**Comment by hanhou** [comment]
Looks like this is indeed what ChatGPT recommends https://chat.openai.com/share/fc9a852f-3b92-4fb5-a4ac-480ea3144780

**Comment by alexpiet** [comment]
Yeah, this is an area where I do not have complete clarity. I think there are a couple things going on right now.
- Yesterday I merged a small PR into `main`, that skipped `production_testing`. I shouldn't have done that because now there are changes that need to get merged from `main` into `production_testing`. I merged it into main thinking that it was just documentation, but it messed up the flow.
- So I just created a PR to merge those updates into `production_testing` (https://github.com/AllenNeuralDynamics/dynamic-foraging-task/pull/98)
- If I look at the PR it has a bunch of commits, just like you saw, and its scary!
<img width="955" alt="Screen Shot 2023-12-08 at 11 07 37 AM" src="https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/7605170/677e16ea-0026-41b9-a607-bbd7191e3ca0">
- But, if you look at the files changed, only the READme is changing, and its from the two commits/1 PR I did yesterday
<img width="1425" alt="Screen Shot 2023-12-08 at 11 08 00 AM" src="https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/7605170/7ab05bfd-064a-4f66-9feb-35200d914e44">
- So I dont really understand why it shows all the commits in this merge. Since all of those commits went into `main` through `production merge`. 
- If we always merge to `production_testing`, not `main` then we can avoid this problem.
- If we always make our updates by branching and then merging back in quickly, then we dont have to worry about the code diverging.
- Maybe we should always merge `main` into `production_testing` after the weekly production merge. The code shouldn't change, but maybe it helps keep the commit history straight? Open to suggestions

**Comment by hanhou** [comment]
Yeah, in sum, the additional commits come from two sources:
1. Commits that were directly merged to `main` without going through `production_testing`. 
    To avoid this, as you said, we should:
   - merge to `production_testing` instead of `main` whenever possible
   - if something is merged directly to `main`, such as a hotfix, merge `main` back to `production_testing` ASAP.

    As long as we follow these two guidelines, we don't need to merge `main` to `production_testing` after each weekly production merge.

3. Those empty "Merge pull request xxx" commits generated automatically by `--no-ff` PR merge, which can be ignored since their diff = 0.
 ----

This read looks very helpful
https://nvie.com/posts/a-successful-git-branching-model/

Their `develop` = our `production_test`, and we kind of combine their `release_branches` and `master` into one, our `main`.

![image](https://github.com/AllenNeuralDynamics/dynamic-foraging-task/assets/24734299/d133ddef-6dad-4a05-a255-7ea3aa535ea0)

Note that their `release` or `master` are merged back to `develop` _**when and only when**_ there are some bugfixes outside `develop` (red circles).

Another point I learn from their workflow is they always create a `freature` branch from `develop`, not `main`. Should we enforce the same?
